File: dajeongkimm/0725/3_14925.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {
    public static void main (String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        int M = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());
        int[][] arr = new int[M][N];
        boolean zero = false;
        for (int i=0;i<M;i++){
            st = new StringTokenizer(br.readLine());
            for (int j=0;j<N;j++){
                arr[i][j] = Integer.parseInt(st.nextToken());

            }
        }
        int[][] dp = new int[M][N];

        int answer = 0;

        // 첫번째 행 초기화
        for (int j=0;j<N;j++){
            if (arr[0][j] == 0) dp[0][j] = 1;

        }
        // 첫번째 열 초기화
        for (int i=0;i<M;i++){
            if (arr[i][0] == 0) dp[i][0] = 1;
        }

        for (int i=1;i<M;i++){
            for (int j=1;j<N;j++){
                if (arr[i][j] == 0){
                    dp[i][j] = Math.min(Math.min(dp[i-1][j-1], dp[i-1][j]),dp[i][j-1]) + 1;
                    answer = Math.max(answer, dp[i][j]);
                }

            }
        }
//        for (int i=0;i<M;i++){
//            for (int j=0;j<N;j++){
//                System.out.print(dp[i][j]+" ");
//            }
//            System.out.println();
//        }
//        System.out.println();

        System.out.println(answer);

    }
}

File: phs7646/0725/3_14925.cpp
#include<iostream>

using namespace std;

bool board[1000][1000];
int dp[1000][1000]; //dp[i][j] : i,j를 우측하단 점으로해서 그릴 수 있는 최대 정사각형
int main() {
    int N,M; cin >> N >> M;
    int answer = 0;
    for(int i = 0;i < N;i++) {
        for(int j = 0;j < M;j++) {
            int a; cin>> a;
            if(a != 0) board[i][j] = true;
            else {
                dp[i][j] = 1;
                answer = 1;
            }
        }
    }
    
    for(int i = 1;i < N;i++) {
        for(int j = 1;j < M;j++) {
            if(board[i][j]) continue;
            //dp[i][j]를 계산
            dp[i][j] = min(dp[i-1][j],dp[i][j-1]);
            dp[i][j] = min(dp[i][j],dp[i-1][j-1]) + 1;
            answer = max(answer,dp[i][j]);
        }
    }
    cout << answer;

    return 0;
}